GH-114 increase coverage in gateway package

makslion opened the pull request:
## What
Coverage change for the gateway package to 90.1%
Extended unit test coverage and created test files for booth controllers. 
No work on the integration right now. 

## Verification
Verification is straightforward - tests are passing on the PR and coe looks neat and covers all the cases it should - PR is verified.

makslion commented:
/hold 
This PR now will wait for #162 and #175 and they will change a lot about the `Gateway` package and tests would need to be done over again

makslion commented:
/label tide/merge-method-squash
/unhold

jasonmadigan commented:
At a skim, this looks OK to me - but it would be good to get another once-over from someone who's familiar with our tests generally as well @mikenairn @david-martin @philbrookes

jasonmadigan commented:
/lgtm

jasonmadigan commented:
/approved

jasonmadigan commented:
/approve

mikenairn commented on pkg/controllers/gateway/test_helper.go:
```diff
@@ -0,0 +1,212 @@
+
+func newTestGatewayPlacer() *TestGatewayPlacer {
+	return &TestGatewayPlacer{}
+}
```
This file is being included as part of the pkg and not a test file, so it's skewing your coverage report. You should probably move it to a `test` package, or fgure out how you can include test helpers for unit tests.

makslion replied:
well, in fact, it bumps coverage up right now :laughing:  but I see the point. 
Why don't we take it even a bit further - I moved all that is not the test cases into the `test` package and once integration tests are done they'd end up in the same `test` package?

mikenairn commented on test/mockInterfaces.go:
```diff
@@ -0,0 +1,302 @@
+
+func NewTestHostService(client client.Client) *TestHostService {
+	return &TestHostService{controlClient: client}
+}
```
Why is this making calls via the client. You are using this in unit tests right? So it shouldn't be making calls to get k8s resources via that API.

There might be a need to mock this interface, but there shouldn't be this much code in it and the values that are going to be returned for each function should be explicit. Your tests could easy fail because of bugs in your test code.

mikenairn replied:
It's being created with a fake client in the unit tests, so can ignore this comment

mikenairn commented on test/mockInterfaces.go:
```diff
@@ -0,0 +1,302 @@
+			Value: "1.1.1.1",
+		},
+	}, nil
+}
```
This ones better, but i think it needs to be more explicit where it's being used what values are going to actually be returned if you are trying to make this into a reusable interface that any test could use where they want to mimic the placement functionality but not actually need to use it.

I had to create a similar one here https://github.com/Kuadrant/multicluster-gateway-controller/pull/237/files#diff-525abdb709ea5f8ab96976a27ec21c0e9545dd47dd82b0e2abff7a475f9672e9R40

mikenairn commented on test/helper.go:
```diff
@@ -0,0 +1,277 @@
+			Namespace: TestNamespace,
+		},
+	}
+}
```
Instead of Get, Build might be a better name for these since you are building a new instance of a resource. Get suggests you are getting it via the API.

I'd also suggest if you want these to be reusable outside of your tests that you add params to a function to let you change appropriate resource values. 

```
func BuildValidTestRequest() ctrl.Request {
	return BuildTestRequest(TestDummyCRName, TestNamespace)
}

func BuildTestRequest(name, ns string) ctrl.Request {
	return ctrl.Request{
		NamespacedName: types.NamespacedName{
			Name:      name,
			Namespace: ns,
		},
	}
}
```

makslion replied:
renamed to `build...`

mikenairn commented on test/helper.go:
```diff
@@ -0,0 +1,277 @@
+	if programmedStatus == v1.ConditionUnknown {
+		message = "current state of the gateway is unknown error %s"
+	}
+	return conditions.SetCondition(gatewayStatus.Conditions, generation, string(v1beta1.GatewayConditionProgrammed), programmedStatus, string(v1beta1.GatewayReasonProgrammed), fmt.Sprintf(message, placed, errorMsg))
```
Why is this here? Is this not code that belongs in the pkg?

makslion replied:
reusing one from the gateway package

mikenairn commented on test/helper.go:
```diff
@@ -0,0 +1,277 @@
+func gatewayClassIsAccepted(gatewayClass *v1beta1.GatewayClass) bool {
+	acceptedCondition := conditions.GetConditionByType(gatewayClass.Status.Conditions, string(v1beta1.GatewayConditionAccepted))
+	return (acceptedCondition != nil && acceptedCondition.Status == v1.ConditionTrue)
+}
```
Same here, what is this in the test package for?

makslion replied:
it is gone now - reusing one from the gateway package

makslion commented on test/mockInterfaces.go:
```diff
@@ -0,0 +1,302 @@
+	panic("unimplemented")
+}
+
+var _ workqueue.RateLimitingInterface = &TestQueue{}
```
resides now in the `eventhandler_test.go`

makslion commented:
resolved the merge conflict and moved things around a bit. The goal is to have the `test` package hold interface implementations and functions that could be reused in other packages. The package-specific helpers now reside at the bottom of the appropriate unit test. Added build flags. 
The conversation remains about what exactly should be covered by unit and integration but this should be as close as we could get right now

mikenairn commented on pkg/controllers/gateway/cluster_eventhandler_test.go:
```diff
@@ -13,24 +13,25 @@ import (
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client/fake"
-	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+	v1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
```
Probably better to leave this to avoid potential confusion with our own local apis

makslion replied:
renamed

mikenairn commented on pkg/controllers/gateway/gateway_controller_test.go:
```diff
@@ -0,0 +1,977 @@
+	}
+}
+
+func TestGatewayReconciler_reconcileDNSEndpoints(t *testing.T) {
```
You need to rebase main, this no longer exists in the gateway controller.

makslion replied:
And a few more tests cases not don't make much sense. Removed

mikenairn commented on test/helper.go:
```diff
@@ -0,0 +1,120 @@
+package test
```
You'll probably want to rebase and make sure this still makes sense here after the e2e tests PR got merged. Also, i think you will need build tags on this

makslion replied:
moved files around. Removed duplicate helper function. Added build tags.

mikenairn commented on pkg/controllers/gateway/params_test.go:
```diff
@@ -55,6 +55,97 @@ func TestGetParams(t *testing.T) {
+						Group:     "",
+						Kind:      "ConfigMap",
+						Name:      "test-params",
+						Namespace: addr(gatewayv1beta1.Namespace("test-ns")),
```
I think you can get rid of the addr function and use testutil.Pointer everywhere?

makslion replied:
missed one! I knew those sneaky calls would get me. Done

mikenairn commented on test/util/helper.go:
```diff
@@ -0,0 +1,118 @@
+const (
+	Domain                 = "thecat.com"
+	ValidTestHostname      = "boop." + Domain
+	FailFetchDNSsubdinain  = "failfetch"
```
FailFetchDNSSubdomain

mikenairn commented on pkg/controllers/gateway/gateway_controller_test.go:
```diff
@@ -0,0 +1,821 @@
+			},
+			verify: func(res ctrl.Result, err error, t *testing.T) {
+				if !reflect.DeepEqual(res, ctrl.Result{}) &&
+					!strings.Contains(err.Error(), "gatewayckasses") &&
```
gatewayckasses?

makslion replied:
the funny part as it was passing with the typo. Fixed

mikenairn commented on pkg/controllers/gateway/gateway_controller_test.go:
```diff
@@ -0,0 +1,821 @@
+				if !reflect.DeepEqual(res, ctrl.Result{}) &&
+					!strings.Contains(err.Error(), "gatewayckasses") &&
+					!strings.Contains(err.Error(), "not found") {
+					t.Errorf("expected to fail finding gateway classm but got err: %s ", err)
```
classm?

makslion replied:
fixed

mikenairn commented:
/lgtm
/approve
